File: library-fe/src/api/books.ts
import { request } from "./client";
import type { Book, BookInput, PaginatedData } from "./types";

export interface ListBooksParams {
  page?: number;
  limit?: number;
  search?: string;
  category?: string;
}

export function listBooks(params: ListBooksParams = {}): Promise<PaginatedData<Book>> {
  return request<PaginatedData<Book>>("/books", { params });
}

export function getBook(id: number): Promise<Book> {
  return request<Book>(`/books/${id}`);
}

export function createBook(input: BookInput): Promise<Book> {
  return request<Book>("/books", { method: "POST", body: input });
}

export function updateBook(id: number, input: Partial<BookInput>): Promise<Book> {
  return request<Book>(`/books/${id}`, { method: "PUT", body: input });
}

export function deleteBook(id: number): Promise<null> {
  return request<null>(`/books/${id}`, { method: "DELETE" });
}

File: library-api/app/api/routes/books.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.schemas.base_response import BaseResponse, PaginatedData
from app.schemas.book import BookCreate, BookResponse, BookUpdate
from app.services.book_service import BookService

router = APIRouter(prefix="/books", tags=["Books"])


@router.get("", response_model=BaseResponse[PaginatedData[BookResponse]])
def list_books(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    search: str | None = Query(None, description="Search by title, author or ISBN"),
    category: str | None = Query(None),
    db: Session = Depends(get_db),
):
    service = BookService(db)
    result = service.list(page=page, limit=limit, search=search, category=category)
    return BaseResponse.success(data=result)


@router.get("/{book_id}", response_model=BaseResponse[BookResponse])
def get_book(book_id: int, db: Session = Depends(get_db)):
    service = BookService(db)
    book = service.get(book_id)
    return BaseResponse.success(data=book)


@router.post("", response_model=BaseResponse[BookResponse], status_code=201)
def create_book(payload: BookCreate, db: Session = Depends(get_db)):
    service = BookService(db)
    book = service.create(payload)
    return BaseResponse.success(data=book, code=201, message="Created")


@router.put("/{book_id}", response_model=BaseResponse[BookResponse])
def update_book(book_id: int, payload: BookUpdate, db: Session = Depends(get_db)):
    service = BookService(db)
    book = service.update(book_id, payload)
    return BaseResponse.success(data=book)


@router.delete("/{book_id}", response_model=BaseResponse[None])
def delete_book(book_id: int, db: Session = Depends(get_db)):
    service = BookService(db)
    service.delete(book_id)
    return BaseResponse.success(data=None, message="Book deleted")
